### ISSUE TITLE
WRF init_ensemble_var.csh perturbations

### ISSUE THREAD

**Issue by timhoar** [NCAR/DART#22]
**Ed note: This issue was originally an email and is being manually ported over to GitHub.**

> I struggle a bit with the DART-WRF tutorial now (https://www.image.ucar.edu/wrfdart/tutorial/):
> **Step 2: Initial conditions** I call 
> ```
> $ ./init_ensemble_var.csh 2017042700
> ```
>  and it returns 
> ```
> $ ncks: ERROR file ../filter_restart_d01.0001 not found. It does not exist on the local filesystem, nor does it match remote filename patterns (e.g., http://foo or foo.bar.edu:file).
> ```
> Where is the file filter_restart_d01.${icnum} written? I don’t seem to have the file and I can’t find any script producing this file.

Regarding the inquiry on the WRF/DART tutorial earlier today, I'm
looking through the latest set of scripts. One thing that seems to be
off is some changes that were made to _init_ensemble_var.csh_ script.
The block of interest is this:
```
       if ( -z add_perts.err ) then
          echo "Perts added to member ${n}"
       else
          echo "ERROR! Non-zero status returned from add_bank_perts.ncl. Check ${RUN_DIR}/advance_temp${n}/add_perts.err."
          cat add_perts.err
          exit
       endif
```

Note that nothing generates a _add_perts.err_ file, so the script ends up
exiting. There is a file generated called _add_perts.out_, but the file
size is not zero. I think this block was added as a way to make sure the
perturbations were successfully added, but maybe this is a placeholder?

**Comment by timhoar** [comment]
If _nclrun3.out_  (which creates _add_perts.out_)  is supposed to create _add_perts.err_ if it fails, then I get the intent of the logic.
I haven't had a chance to look at this yet.

**Comment by jlsucar** [comment]
`add_bank_perts.ncl` does reference `add_perts.err` through the following strange mechanism:

```
  err = NhlGetErrorObjectId()
  setvalues err
    "errFileName" : "add_perts.err"
  end setvalues
```

It seems from the documentation that this can be used to set the error out location https://www.ncl.ucar.edu/Document/Functions/Built-in/NhlGetErrorObjectId.shtml

Thus, I think the -z logic above is correct. Ben and I had gone through this script and logic a few months ago and it worked correctly at that time.

**Comment by timhoar** [comment]
FYI:  Actually ... `-e` checks for existence,    `-z` checks for zero size ...

**Comment by g-romine** [comment]
Jeff,

you can make a toy code for this and see that it does not work as expected. For example:

#!/bin/csh

       set n = 1
       if ( -z add_perts.err ) then
          echo "Perts added to member ${n}"
       else
          echo "ERROR! Non-zero status returned from add_bank_perts.ncl. Check /advance_temp${n}/add_perts.err."
          cat add_perts.err
          exit
       endif

If you run this script, you will see that you get the error message and the script exits, which in the full script means it doesn't complete running the tasks needed. If you touch the file add_perts.err (so, it is a file that is present but of size zero), then the logic test passes and the script continues. I don't know how you get the .err file to be created, I've never encountered one, but the more important thing is whether the script ran correctly, as in did it actually add a perturbation to the file, and checking that can be done as follows (new and modified lines in bold):

#!/bin/csh

       set n = 1
       grep 'perts added' add_perts.out > add_perts.err
       if (! -z add_perts.err ) then
          echo "Perts added to member ${n}"
       else
          echo "ERROR! Non-zero status returned from add_bank_perts.ncl. Check /advance_temp${n}/add_perts.err."
          cat add_perts.err
          exit
       endif

Glen

**Comment by jlsucar** [comment]
Ah, yes, thank you.

I just mocked up how ncl will behave here:

```
begin

  err = NhlGetErrorObjectId()
  setvalues err
    "errFileName" : "add_perts.err"
  end setvalues

  print("hello")
end
```

```
module load ncl
ncl test.ncl
ls -alrt add_perts.err
```

And the result is:

```
-rw-r--r--  1 jsteward ncar      0 Jun 18 12:57 add_perts.err
```

So, it does exist, and it is a 0-length file. The (somewhat strange) logic should be correct.

I think we should add a comment at that point of the script saying what the heck is going on.

**Comment by g-romine** [comment]
Well, above it wiped out my bold formatting, but the two lines changed in the 2nd example are:

   grep 'perts added' add_perts.out > add_perts.err
   if (! -z add_perts.err ) then

**Comment by jlsucar** [comment]
Hi Glen,

If you'd like to jump on the software standup, we could discuss this issue
briefly.

Best wishes,

Jeff

**Comment by timhoar** [comment]
OK - so now you two are talking across purposes ... 

if _add_bank_perts.ncl_  is supposed to create `add_perts.err` **_only if there is an error_**,
then checking for the existence of the file is the correct course of action.

If it _always_ creates the file, then checking if it is zero length is the correct action.

Glen indicates he can run and the file is not created, so it seems to me the we should be replacing
```
if ( -z add_perts.err ) then
```
with
```
if ( -e add_perts.err ) then
```

**Comment by timhoar** [comment]
To be complete, the logic of the test also has to be inverted:
```
       if ( -e add_perts.err ) then
          echo "ERROR! Non-zero status returned from add_bank_perts.ncl. Check ${RUN_DIR}/advance_temp${n}/add_perts.err."
          cat add_perts.err
          exit 1
      else
          echo "Perts added to member ${n}"
       endif
```

to me, this is pretty clear.
It is a bit mysterious where `add_perts.err` comes from, so a comment to that effect would seem warrented.

**Comment by jlsucar** [comment]
Yeah, we were both, at the same time, writing up about related but slightly different issues. Note that neither Glen nor I saw what the other was writing above.

I just verified the logic as follows.

If the `./shell_scripts/add_bank_perts.ncl` script executes, it will create an error file no matter what. Whether that file is zero length (no error) or non-zero length (there was an error) is the main issue here.

So I maintain the logic is correct as written, and the non-zero length `add_perts.err` file should be deleted and re-run. However, there may be another issue with the logic of `add_bank_perts.ncl` not being called, or being called correctly.

Proof:

Outside of any script context, run 

`ncl add_bank_perts.ncl`

There is a `add_perts.err` file, non-zero length, as expected. The contents are:

```
fatal:Variable (MEM_NUM) is undefined
^Mfatal:["Execute.c":8637]:Execute: Error occurred at or near line 16 in file shell_scripts/add_bank_perts.ncl
```

Now, if you run the above test.ncl:

```
begin

  err = NhlGetErrorObjectId()
  setvalues err
    "errFileName" : "add_perts.err"
  end setvalues

  print("hello")
end
```

A zero-length file add_perts.err file is created.

Now, again, the bigger issue is whether `add_pert_bank.ncl` ever gets called, or should get called.

In short I think the issue isn't with the `-z` aspect at all.

**Comment by g-romine** [comment]
The .err file doesn't come natively, it can only be produced by modifying the ncl code (and the code we include the in the tutorial doesn't have this modification). Still, this is fairly easy to fix - we can just add those lines to the ncl code (add_bank_perts.ncl) in the rundir, and then it works with the shell script logic as is. I'm not sure what happens if the ncl script is interrupted, but once it returns, if a problem occurred during execution something gets written to the file, and it is created. Failure of the file to to be created at all can happen with bad or missing command line arguments (or ncl executable not being in the path), and then no file is created and the script exits. A more informative error message could be provided here, but the intent seems clear. So, let's simply update the ncl script for now.

**Comment by jlsucar** [comment]
Thanks Glen, I think we are on the same page, except I have a question about which ncl script is being run. 

This does have the `add_perts.err` code:

https://github.com/NCAR/DART/blob/Manhattan/models/wrf/shell_scripts/add_bank_perts.ncl

Which version is being used by the tutorial?

**Comment by g-romine** [comment]
Jeff - that file is not the same as the one in the wrf/dart tutorial bundle (see step 1 on https://www.image.ucar.edu/wrfdart/tutorial/). Sounds like it should be removed from the bundle so we don't have the same file in multiple locations.

**Comment by jlsucar** [comment]
Great - I think the issue is solved. The bundle on that page has outdated code versus the master. @timhoar what do we need to do to update that bundle, and hopefully do so in a way that we won't be out of sync again?

**Comment by fossell** [comment]
Hi All - It sounds like this is resolved and was just a matter of using the updated script?  Please let me know if there is still an issue and something I can help with.  Apologies for being late to the party, long day of dealing with issues on another github project.

**Comment by timhoar** [comment]
Consolidate the tutorial/README and index.html ... put in a redirect to the result.
